File: 12_simd_intrinsics/problems/00_vector_fma/vector_fma.f90
! Fortran には C/C++ のベクトル型拡張 (vector_size) に相当する機能が無い.
! そこで, 同じ y(i) = a(i) * b(i) + c(i) の積和 (fma) をループで書き,
! !$omp simd で SIMD 化する.
subroutine vector_fma(n, a, b, c, y)
  implicit none
  integer(8), intent(in) :: n
  real(8), intent(in) :: a(n), b(n), c(n)
  real(8), intent(out) :: y(n)
  integer(8) :: i
  ! TODO: 下の do ループの直前に !$omp simd を1行追加し, このループをSIMD化せよ.
  do i = 1, n
     y(i) = a(i) * b(i) + c(i)
  end do
end subroutine vector_fma

File: 12_simd_intrinsics/problems/00_vector_fma/vector_fma.cpp
#include <cstdio>

/* double 8 つ分 (8 × 8 = 64 バイト = 512 bit) のベクトル型 */
typedef double doublev __attribute__((vector_size(64)));

/* a, b, c はいずれも double 8 つ分のベクトル型.
   a*b+c を要素ごとの演算 (fma) としてベクトルのまま計算して返す. */
doublev vector_fma(doublev a, doublev b, doublev c) {
  // TODO: a * b + c を返す1行を書け.
}
